Feature/provider

Lee-Yoonjoo opened the pull request:
Added provider patter for WidgetList, IconList and ExpandableList

Lee-Yoonjoo committed:
expandablelist feature will be closed.

Lee-Yoonjoo committed:
Add provider library

Lee-Yoonjoo committed:
Add provider patter for ListItem

Lee-Yoonjoo committed:
Add provideer pattern for icon list and expandable list

mkoechig commented:
login.dart file is obsolete and should be deleted.
gitflow.dart consists of text only and should not be a dart file. The notes can be moved into README.md instead and also make use of the markdown formatting.

Lee-Yoonjoo committed:
created icondetail.dart

Lee-Yoonjoo committed:
Update feature/provider on feature/iconitemdetail

Lee-Yoonjoo committed:
Since Provider was used, repo constructor has been removed

Lee-Yoonjoo committed:
Set routes for detail page

Lee-Yoonjoo committed:
Corrected a mistake. on ExpandableList, provider should have called loadIntItems.

Lee-Yoonjoo committed:
Added IconDetail with Head Section

Lee-Yoonjoo committed:
Considering how to get mock data

Lee-Yoonjoo committed:
Considering how to get mock data

Lee-Yoonjoo committed:
Removed login.dart, move notes to README.md

Lee-Yoonjoo committed:
Change some design. Added padding and rounded shape on list tile.

mkoechig commented on lib/providers/provider.dart:
```diff
@@ -0,0 +1,29 @@
+import 'package:mixeditems/models/listitem.dart';
+import 'package:mixeditems/repository/repo.dart';
+
+class CustomProvider extends ChangeNotifier {
```
CustomProvider is not a clear name for this. Use more explicit naming, like, what is this provider providing?

Lee-Yoonjoo replied:
I think people can use different provider by data model
I will take IconListProvider for this project and handle other data on another Provider or another way. 

I have checked his provider architecture sample. https://github.com/brianegan/flutter_architecture_samples/blob/master/change_notifier_provider/lib/todo_list_model.dart

Lee-Yoonjoo replied:
I am not so sure for now, which is the breakpoint for multiproviders .. Mit anderen Worten when we will need to create new Provider...

mkoechig replied:
As a rule of thumb: when the class get's too big :D
500 Lines of code is often considered a soft limit.

mkoechig commented on lib/providers/provider.dart:
```diff
@@ -0,0 +1,29 @@
+import 'package:mixeditems/repository/repo.dart';
+
+class CustomProvider extends ChangeNotifier {
+  Repo _data = Repo();
```
It rather is a data source than actual data

Lee-Yoonjoo replied:
I think that it will be better with _repo.

mkoechig commented on lib/providers/provider.dart:
```diff
@@ -0,0 +1,29 @@
+
+  List<int> get intItems => _intItems;
+
+  loadListItems() {
+    _listItems = _data.getListItems();
+  }
+
+  loadIconItems() {
+    _iconItems = _data.getIconListItems();
+  }
+
+  loadIntItems() {
+    _intItems = _data.getMockData();
+  }
```
Ignore this for now... placeholder^^

These should convert data to widgets. See Repo class for details.

Lee-Yoonjoo replied:
I understand. I tried to keep them clear !

mkoechig commented on lib/widgets/home.dart:
```diff
@@ -0,0 +1,34 @@
+import 'package:mixeditems/providers/provider.dart';
+import 'package:provider/provider.dart';
+
+class Home extends StatelessWidget {
```
Dart has a coding guideline to have a keyed constructor in any widget: https://dart-lang.github.io/linter/lints/use_key_in_widget_constructors.html

Lee-Yoonjoo replied:
I have changed it! :)

mkoechig commented on lib/widgets/home.dart:
```diff
@@ -0,0 +1,34 @@
+import 'package:mixeditems/providers/provider.dart';
+import 'package:provider/provider.dart';
+
+class Home extends StatelessWidget {
```
I would suffix Widgets with a "...Widget" to make things clear

Lee-Yoonjoo replied:
Oh Okay, I changed others classes names which extends StatefulWidget or StatelessWidget.

mkoechig reviewed (commented):
A repository should only return models, not widgets.
Widgets are view entities (presentation layer) and don't belong into a repository (domain layer).
It should be a widget's or it's provider's responsibility to take the data model and create a view/widget with the information provided within.

I will research some common provider implementations and update this later on with more information on how to use it.

mkoechig commented:
This branch should not merge to master, but develop instead

Lee-Yoonjoo committed:
Changed names properly

Lee-Yoonjoo committed:
Changed names properly_2

Lee-Yoonjoo committed:
Removed Abstract class and make each section extends StatelessWidget

Lee-Yoonjoo committed:
The corrections have been reflected.
1. Changed names to right names.
2. Clarified the provider roles.
3. Tried to keep Widgets and logics separate.

mkoechig commented on lib/repository/repo.dart:
```diff
@@ -1,34 +1,35 @@
     return [1, 2, 3, 4, 5];
   }
+
+  List<IconDetail> getDataUsageList (){
```
In this class you didn't hit the auto format, did you?

Lee-Yoonjoo replied:
I am not sure so I just did Auto format :)

mkoechig commented on lib/repository/widget_placeholder.dart:
```diff
@@ -0,0 +1,35 @@
+import 'package:mixeditems/sections/icon_detail/usage_section.dart';
+import 'package:mixeditems/sections/icon_detail/version_section.dart';
+
+class WidgetPlaceHolder {
```
Why placeholder? This seems to be a 'factory', as it's creating Widgets.

mkoechig commented on lib/repository/widget_placeholder.dart:
```diff
@@ -0,0 +1,35 @@
+import 'package:mixeditems/sections/icon_detail/usage_section.dart';
+import 'package:mixeditems/sections/icon_detail/version_section.dart';
+
+class WidgetPlaceHolder {
```
This class is placed inside the repository package. But it is creating Widgets and therefore views. Repositories should only handle data and not handle nor create Views.

Lee-Yoonjoo replied:
I will move them in its widget class.

mkoechig commented on lib/repository/widget_placeholder.dart:
```diff
@@ -0,0 +1,35 @@
+    ];
+  }
+
+}
```
This class also has not been services by the auto formatter

Lee-Yoonjoo replied:
I will make sure

mkoechig commented on lib/screens/icon_detail_widget.dart:
```diff
@@ -0,0 +1,39 @@
+  }
+
+  Widget _buildListView(BuildContext context) {
+    WidgetPlaceHolder _widgetPlaceHolder = WidgetPlaceHolder();
```
This should be a property of the class, instead of being newly created everytime this function is called

Lee-Yoonjoo replied:
I see I will move them in the widget class.

mkoechig replied:
That's not what I meant, but it's okay for now. It's not the scope of this branch.

Lee-Yoonjoo committed:
The corrections have been reflected.
1. Removed unnecessary placeholder
2. Applied Auto format for each class
